src/Transform/ONNX/ONNXDimAnalysis.cpp

```diff
@@ -0,0 +1,588 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===-------- ONNXDimAnalysis.cpp - ONNX Dimension Analysis ---------------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file implements an analysis on unknown dimensions in ONNX ops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+#include "llvm/Support/Debug.h"
+
+#include "src/Dialect/ONNX/DialectBuilder.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Dialect/ONNX/ONNXOpsHelper.hpp"
+#include "src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp"
+#include "src/Pass/Passes.hpp"
+#include "src/Support/TypeUtilities.hpp"
+#include "src/Transform/ONNX/ONNXDimAnalysis.hpp"
+
+#define DEBUG_TYPE "dim_analysis"
+
+using namespace mlir;
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Helper functions for dimension analysis.
+//===----------------------------------------------------------------------===//
+
+/// Check if two sets are interleaved or not.
+bool isInterleaved(const onnx_mlir::DimAnalysis::DimSetT &lhs,
+    const onnx_mlir::DimAnalysis::DimSetT &rhs) {
+  // Call `contains` on the smaller set for better performance.
+  if (lhs.size() < rhs.size()) {
+    for (auto &ti : lhs) {
+      if (rhs.contains(ti)) {
+        return true;
+      }
+    }
+  } else {
+    for (auto &ti : rhs) {
+      if (lhs.contains(ti)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+/// Given a QuestionMarkIndexExpr representing an unknown dimension, find the
+/// same unknown dimensions in the inputs.
+void findAndAddSameDim(const onnx_mlir::QuestionmarkIndexExpr &qmOuputIE,
+    mlir::ValueRange operands, onnx_mlir::DimAnalysis::DimSetT &sameDims) {
+  int64_t outputQMValue = qmOuputIE.getQuestionmark();
```

Comment:
Ideally this call would not be needed, because we only query the index expression directly for the info we need. See below.

Reply:
Could you elaborate a bit? I don't see "below" things, maybe you forgot pushing them ...

Reply:
To the extend possible, we should not need the `getQuestionmark()` call.
We should have
```
bool qmOutputIE.specificQuestionmark() 
bool qmOutputIE.sameQuestionmark(iIR)
```
where we ask index expression if an index expression question mark is specific (<-1) or not (==-1), and if they are the same (same specific number).

If we were to change the representation, it would not impact the code here.

Reply:
Thanks for th explaination! I addressed this just now.